File: android/src/main/java/com/salvedb/BackgroundScheduleDecision.kt
package com.salvedb

sealed interface BackgroundScheduleDecision {
  data object Cancel : BackgroundScheduleDecision
  data class Enqueue(
    val intervalMs: Long,
    val requiresNetwork: Boolean,
    val requiresCharging: Boolean,
  ) : BackgroundScheduleDecision
}

fun backgroundScheduleDecision(
  hasConfig: Boolean,
  minimumIntervalMs: Double,
  requiresNetwork: Boolean,
  requiresCharging: Boolean,
  floorMs: Long,
): BackgroundScheduleDecision {
  if (!hasConfig) return BackgroundScheduleDecision.Cancel
  return BackgroundScheduleDecision.Enqueue(
    intervalMs = minimumIntervalMs.toLong().coerceAtLeast(floorMs),
    requiresNetwork = requiresNetwork,
    requiresCharging = requiresCharging,
  )
}

File: ios/Sync/BackgroundScheduleDecision.swift
import Foundation

enum BackgroundScheduleDecision: Equatable {
  case cancel
  case enqueue(secondsFromNow: TimeInterval, requiresNetwork: Bool, requiresCharging: Bool)
}

func backgroundScheduleDecision(
  hasConfig: Bool,
  minimumIntervalMs: Double,
  requiresNetwork: Bool,
  requiresCharging: Bool
) -> BackgroundScheduleDecision {
  guard hasConfig else { return .cancel }
  return .enqueue(
    secondsFromNow: minimumIntervalMs / 1000,
    requiresNetwork: requiresNetwork,
    requiresCharging: requiresCharging
  )
}
